### PR TITLE
Remove -Zfuel.

### PR DESCRIPTION
**Pull request by cjgillot**
I'm not sure this feature is used. I only found 2 references in a google search, both referring to its introduction.

Meanwhile, it's a global mutable state, untracked by incremental compilation, so incompatible with it.

### CODE REVIEW

**Comment by JakobDegen** [comment]
I agree we should remove it right now, primarily because it's usefulness is significantly diminished by inconsistent use. At some point in the future we should switch to more structured transformations (as opposed to writing to `&mut`s) in which case this can be brought back and we can enforce consistent usage

**Comment by wesleywiser** [comment]
Similar to @tmiasko, I've also found this useful in the past. The point that this is inconsistently used is true, but wouldn't it be better to fix that than to remove it entirely?

**Comment by cjgillot** [comment]
We now have a more reliable way to disable optimizations: `-Zmir-enable-passes`. I recommend using that one for debugging.

**Review by wesleywiser — APPROVED** [review]
r=me with a corresponding PR to adjust rustc-dev-guide

**Comment by bjorn3** [comment]
`-Zfuel` is entirely unreliable with parallel rustc, right? The order in which functions get mir optimizations running is non-deterministic and thus which function sees the last fuel consumed.

**Comment by nnethercote** [comment]
FWIW, I just wrote a patch to refactor the `OptimizationFuel` type (because it's a bit messy) and then I just discovered this PR...

**Comment by Dylan-DPC** [comment]
@cjgillot any updates on this? thanks

**Comment by cjgillot** [comment]
> r=me with a corresponding PR to adjust rustc-dev-guide

Here : https://github.com/rust-lang/rustc-dev-guide/pull/2032

File: compiler/rustc_mir_transform/src/instsimplify.rs [hunk 1]
```diff
@@ -160,36 +139,24 @@ impl<'tcx> InstSimplifyContext<'tcx, '_> {
...
         if let Rvalue::Len(ref place) = *rvalue {
             let place_ty = place.ty(self.local_decls, self.tcx).ty;
             if let ty::Array(_, len) = *place_ty.kind() {
-                if !self.should_simplify(source_info, rvalue) {
-                    return;
-                }
-
                 let const_ = Const::from_ty_const(len, self.tcx.types.usize, self.tcx);
-                let constant = ConstOperand { span: source_info.span, const_, user_ty: None };
+                let constant = ConstOperand { span: DUMMY_SP, const_, user_ty: None };
```
Comment (wesleywiser) [inline compiler/rustc_mir_transform/src/instsimplify.rs:147]: Why are we losing the spans?

Comment (cjgillot) [reply compiler/rustc_mir_transform/src/instsimplify.rs:147]: We usually ignore spans for constants in optimisations, as they may only be used in the unfortunate case of post-mono errors, which must not depend on optimisations running.

**Comment by nnethercote** [comment]
It will be good to get this merged. I have some follow-up cleanups that rely on `-Zfuel` being removed :)

**Review by dianqk — APPROVED** [review]
It will be good to get this merged as well. I don’t use `-Zfuel`, and the reliability of this option still depends on whether we remember to call `consider_optimizing`.

**Comment by DianQK** [comment]
♥️
@bors r=wesleywiser,DianQK

**Comment by DianQK** [comment]
🧐

**Pull request merged**